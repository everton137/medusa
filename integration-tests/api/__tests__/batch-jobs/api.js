const path = require("path")

const setupServer = require("../../../helpers/setup-server")
const { useApi } = require("../../../helpers/use-api")
const { initDb, useDb } = require("../../../helpers/use-db")

const adminSeeder = require("../../helpers/admin-seeder")
const userSeeder = require("../../helpers/user-seeder")
const { simpleBatchJobFactory } = require("../../factories")

jest.setTimeout(50000)

const adminReqConfig = {
  headers: {
    Authorization: "Bearer test_token",
  },
}

describe("/admin/batch", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    dbConnection = await initDb({ cwd })
    medusaProcess = await setupServer({ cwd, verbose: false })
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  describe("POST /admin/batch/:id/complete", () => {
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        await userSeeder(dbConnection)

        await simpleBatchJobFactory(dbConnection, {
          id: "job_1",
          type: "batch_1",
          created_by: "admin_user",
          status: "created",
        })
        await simpleBatchJobFactory(dbConnection, {
          id: "job_2",
          type: "batch_1",
          status: "awaiting_confirmation",
          created_by: "admin_user",
        })
        await simpleBatchJobFactory(dbConnection, {
          id: "job_3",
          type: "batch_1",
          status: "awaiting_confirmation",
          created_by: "member-user",
        })
      } catch (err) {
        console.log(err)
        throw err
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("Completes batch job created by the user", async () => {
      const api = useApi()

      const jobId = "job_2"

      const response = await api.post(
        `/admin/batch/${jobId}/complete`,
        {},
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data.batch_job).toMatchSnapshot({
        created_at: expect.any(String),
        updated_at: expect.any(String),
        completed_at: expect.any(String),
        status: "completed",
      })
    })

    it("Fails to complete a batch job created by a different user", async () => {
      const api = useApi()

      const jobId = "job_3"

      api
        .post(`/admin/batch/${jobId}/complete`, {}, adminReqConfig)
        .catch((err) => {
          expect(err.response.status).toEqual(400)
          expect(err.response.data.type).toEqual("not_allowed")
          expect(err.response.data.message).toEqual(
            "Cannot complete batch jobs created by other users"
          )
        })
    })

    it("Fails to complete a batch job not awaiting completion", async () => {
      const api = useApi()

      const jobId = "job_1"

      api
        .post(`/admin/batch/${jobId}/complete`, {}, adminReqConfig)
        .catch((err) => {
          expect(err.response.status).toEqual(400)
          expect(err.response.data.type).toEqual("invalid_data")
          expect(err.response.data.message).toEqual(
            `Cannot complete a batch job with status "created"`
          )
        })
    })
  })

  describe("GET /admin/batch", () => {
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        await userSeeder(dbConnection)

        await simpleBatchJobFactory(dbConnection, {
          id: "job_1",
          type: "batch_1",
          created_by: "admin_user",
        })
        await simpleBatchJobFactory(dbConnection, {
          id: "job_2",
          type: "batch_2",
          created_by: "admin_user",
        })
        await simpleBatchJobFactory(dbConnection, {
          id: "job_3",
          type: "batch_2",
          created_by: "admin_user",
        })
        await simpleBatchJobFactory(dbConnection, {
          id: "job_4",
          type: "batch_1",
          created_by: "member-user",
        })
      } catch (err) {
        console.log(err)
        throw err
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("lists batch jobs created by the user", async () => {
      const api = useApi()

      const response = await api.get("/admin/batch", adminReqConfig)

      expect(response.status).toEqual(200)
      expect(response.data.batch_jobs.length).toEqual(3)
      expect(response.data).toMatchSnapshot({
        batch_jobs: [
          {
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
          {
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
          {
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
        ],
      })
    })
  })
})
