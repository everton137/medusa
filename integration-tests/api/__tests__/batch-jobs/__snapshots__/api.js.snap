// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/admin/batch GET /admin/batch lists batch jobs created by the user 1`] = `
Object {
  "batch_jobs": Array [
    Object {
      "context": Object {},
      "created_at": Any<String>,
      "created_by": "admin_user",
      "deleted_at": null,
      "id": "job_3",
      "result": null,
      "status": "created",
      "type": "batch_2",
      "updated_at": Any<String>,
    },
    Object {
      "context": Object {},
      "created_at": Any<String>,
      "created_by": "admin_user",
      "deleted_at": null,
      "id": "job_2",
      "result": null,
      "status": "created",
      "type": "batch_2",
      "updated_at": Any<String>,
    },
    Object {
      "context": Object {},
      "created_at": Any<String>,
      "created_by": "admin_user",
      "deleted_at": null,
      "id": "job_1",
      "result": null,
      "status": "created",
      "type": "batch_1",
      "updated_at": Any<String>,
    },
  ],
  "count": 3,
  "limit": 10,
  "offset": 0,
}
`;

exports[`/admin/batch POST /admin/batch/:id Updates batch job 1`] = `
Object {
  "awaiting_confirmation_at": null,
  "cancelled_at": null,
  "completed_at": null,
  "context": Object {},
  "created_at": Any<String>,
  "created_by": "admin_user",
  "deleted_at": null,
  "id": Any<String>,
  "processing_at": null,
  "result": null,
  "status": "created",
  "type": "batch_2",
  "updated_at": Any<String>,
}
`;

exports[`/admin/batch POST /admin/batch/:id Updates batch job with status sets time of update 1`] = `
Object {
  "awaiting_confirmation_at": null,
  "cancelled_at": null,
  "completed_at": null,
  "context": Object {},
  "created_at": Any<String>,
  "created_by": "admin_user",
  "deleted_at": null,
  "id": Any<String>,
  "processing_at": Any<String>,
  "result": null,
  "status": "processing",
  "type": "batch_1",
  "updated_at": Any<String>,
}
`;

exports[`/admin/batch POST /admin/batch/:id/complete Completes batch job created by the user 1`] = `
Object {
  "awaiting_confirmation_at": null,
  "cancelled_at": null,
  "completed_at": Any<String>,
  "context": Object {},
  "created_at": Any<String>,
  "created_by": "admin_user",
  "deleted_at": null,
  "id": "job_2",
  "processing_at": null,
  "result": null,
  "status": "completed",
  "type": "batch_1",
  "updated_at": Any<String>,
}
`;
